README.md
# libnmap 

API to access nmap from node.js

[![npm](https://img.shields.io/npm/v/libnmap.svg)](https://npmjs.com/package/libnmap) [![Build Status](https://travis-ci.org/jas-/node-libnmap.png?branch=master)](https://travis-ci.org/jas-/node-libnmap) [![Dependencies](https://img.shields.io/david/jas-/node-libnmap.svg)](https://david-dm.org/jas-/node-libnmap) ![Downloads](https://img.shields.io/npm/dm/libnmap.svg) [![Known Vulnerabilities](https://snyk.io/test/github/jas-/node-libnmap/badge.svg?targetFile=package.json)](https://snyk.io/test/github/jas-/node-libnmap?targetFile=package.json)


## install ##

To install `npm install libnmap`

## methods ##

* `scan`      Performs scan given available range & optional port
* `discover`  Retrieves list of online network neighbors

## options ##

* `nmap`      {String}    Path to NMAP binary
* `verbose`   {Boolean}   Turn on verbosity during scan(s)
* `ports`     {String}    Range of ports to scan
* `range`     {Array}     An array of hostnames/ipv4/ipv6, CIDR or ranges
* `timeout`   {Number}    Number of minutes to wait for host/port response
* `blocksize` {Number}    Number of hosts per network scanning block
* `threshold` {Number}    Max number of  spawned process
* `flags`     {Array}     Array of flags for .spawn()
* `udp`       {Boolean}   UDP scan mode enabled
* `json`      {Boolean}   JSON object as output, false produces XML

## tests ##

To test `npm test`

## examples ##

A default usage example. For more advanced and possible options please
see [here](https://github.com/jas-/node-libnmap/tree/master/examples) or simply
look in the included `examples/` folder.

### scan ###

The example show shows the types of host ranges supported. In this example the
default IANA range of reserved ports is scanned per host in each range (1024).

```javascript
const nmap = require('libnmap');
const opts = {
  range: [
    'scanme.nmap.org',
    '10.0.2.0/25',
    '192.168.10.80-120',
    'fe80::42:acff:fe11:fd4e/64'
  ]
};

nmap.scan(opts, function(err, report) {
  if (err) throw new Error(err);

  for (let item in report) {
    console.log(JSON.stringify(report[item]));
  }
});
```

### discover ###

The discover method requires nodejs < `v0.11.2` and can be used to aquire
information about neighbors per network interface.

```javascript
const nmap = require('libnmap');

nmap.discover(function(err, report) {
  if (err) throw new Error(err);

  for (let item in report) {
    console.log(JSON.stringify(report[item]));
  }
});
```

### sample reports ###

To see some output examples please take a look at the [json (default) & xml reports](https://gist.github.com/jas-/23e2a32110562388bef5).

## contributing ##

Contributions are welcome & appreciated. Refer to the [contributing document](https://github.com/jas-/node-libnmap/blob/master/CONTRIBUTING.md)
to help facilitate pull requests.

## license ##

This software is licensed under the [MIT License](https://github.com/jas-/node-libnmap/blob/master/LICENSE).

Copyright Jason Gerfen, 2013-2018.


CHANGELOG
commit d50f2485836d9a6a2eb449c68572b9aa606b7ef2
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Oct 19 13:05:43 2018 +0000

    v0.4.15: Version bump & depedency migration to patched cidr-js

commit 743fc80c0a259848c14663f596a771d80303aca5
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Sep 16 20:28:48 2018 +0000

    v0.4.14: CHANGELOG updated

commit 6103871b1ceae361958fa194a24731f5754ba3f3
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Sep 16 15:08:00 2018 +0000

    v0.4.14: Removed conditional for linux pre install

commit 02204990f19738c7e6b33c6696f3a1a17e675137
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Sep 16 15:03:51 2018 +0000

    v0.4.14: Removed conditional for linux pre install

commit 6e797385f80d20fcba1771a49d1519f002647afd
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Sep 16 14:54:35 2018 +0000

    v0.4.14: Fix for .match()

commit f51515317fd14a2d752eb4d1cb9339ef40cf38a3
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sat Sep 15 14:15:53 2018 +0000

    v0.4.14: Fix for use of .match()

commit ba26dafeaf232539050ec5d7ca03cf05aafbba17
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sat Sep 15 00:20:32 2018 +0000

    v0.4.14: Travis-CI: OSX fix candidate #004

commit 8bcad169d3254b72e050f232130d5050dd554fb3
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sat Sep 15 00:09:51 2018 +0000

    v0.4.14: Travis-CI: OSX fix candidate #003

commit 355c2ce6f34a50bbb64128e4c6612db6b48523d4
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sat Sep 15 00:00:48 2018 +0000

    v0.4.14: Travis-CI: OSX fix candidate #002

commit 1928f849586a627193bca082a7e070844af084c3
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Sep 14 23:49:16 2018 +0000

    v0.4.14: Travis-CI: OSX fix candidate #001

commit 42354941745c92da5132ed033d700d24503c66c9
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Sep 14 11:04:16 2018 +0000

    v0.4.14: Fix for Travis-CI (OSX)

commit 9270388d31400ccaecadea8143906a0f13d8100a
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Sep 14 03:31:28 2018 +0000

    Updated .gitignore to remove folder from npm publish

commit 7cfcd39525755194d905c8b9caf66e52c97d7736
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Sep 14 03:24:52 2018 +0000

    v0.4.13: 20180914
     - Potential candidate for issues #53 & #54
     - Removed validation test harness
     - Moved away from .test() to .match() for opts.ports validation

commit 3640d0b6533c9a439b3bdc88f7e49cf971bfbfad
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Aug 26 14:48:11 2018 +0000

    Updated changelog & version number

commit 7d624968036ca23604cdbd8b84cbf033fb5e393a
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Aug 26 14:33:12 2018 +0000

    Implemented package locking

commit 92eac0ede90113aa9ed9fd460f041b2c379132c7
Author: Jas- <jason.gerfen@gmail.com>
Date:   Sun Aug 26 14:31:04 2018 +0000

    v0.4.13: Version bump

commit c5f639d04b8bbb9f967df5909ffc451320b3db2f
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Apr 13 06:12:06 2018 +0000

    CHANGELOG update

commit 52421c9ec95e5cf616f2b49033658865a8466025
Author: Jas- <jason.gerfen@gmail.com>
Date:   Fri Apr 13 06:11:15 2018 +0000

    Version bump in package.json

commit 1b6fcc1e4871ed35ec97450023b2df0399a49527
Merge: efce9d5 41c2922
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 13:53:38 2018 +0000

    1

commit efce9d5c23560d6672d4b19ab95ef11c19f150a3
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:12:53 2018 +0000

    v0.4.11 - Staged version bump
    
    Addresses unhandled exceptions with use of error domain
    Creates module based libraries
    Moves further towards ES6

commit 41c2922958b0bba15a232d2999577f434ae43028
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 11:04:33 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part nine

commit d916d31b3e79dd9516872cb29b8fa2de51bfeb6a
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 10:58:28 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part eight

commit 7816c56ab1a8c74c79ad49919b0b707be626aad1
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 10:19:13 2018 +0000

    Moved anonymous functions to ES6 notation

commit 722fea8cc93ad30408d1ee2a49945d73dd5978c6
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 08:54:07 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part seven

commit 2739e883aeb3989b64c621d3c8a08f7c8bf6a27b
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 08:40:48 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part six

commit e4b18caca0dae50efb082d0b32c9e3245cd7352e
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 08:32:05 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part five

commit 6222edbf6d0ef6be1e22bc8834e1721387a6d061
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:56:57 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part four

commit 214d78df20670504c814cc93333cba20b92a8305
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:44:06 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part tre

commit eb55bb93bb563169c5804866fa19377aea8cef4d
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:19:22 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx; part duex

commit 5a10ca13d341467e72d4b31d4aa274cb95c31c2a
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:12:53 2018 +0000

    Attempts to fix nvm version for travis-ci in linux/osx

commit bd1f79f8f04eaabbbd284167d53751f39317560f
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 07:00:27 2018 +0000

    Added iteration to handle those adapters using multiple IP addresses per RFC 7934

commit 8c001d241db59a191e9bd7f2b05c0e33863e7d57
Author: Jas- <jason.gerfen@gmail.com>
Date:   Wed Mar 21 06:52:29 2018 +0000


lib/libnmap-es6.js
/*!
 * libnmap
 * Copyright(c) 2013-2017 Jason Gerfen <jason.gerfen@gmail.com>
 * License: MIT
 */

const version = 'v0.2.33'
    , fs = require('fs')
    , os = require('os')
    , async = require('async')
    , hasbin = require('hasbin')
    , xml2js = require('xml2js')
    , ip = require('ip-address')
    , merge = require('deepmerge')
    , caller = require('caller-id')
    , netmask = require('netmask').Netmask
    , proc = require('child_process').exec
    , v6 = ip.Address6
    , nmap = function(options, fn) {
console.log(typeof v6)
console.log(v6)
  'use strict'

  /**
   * @object defaults
   * @abstract Default set of options
   *
   * @param {String} nmap - Path to NMAP binary
   * @param {Boolean} verbose - Turn on verbosity during scan(s)
   * @param {String} ports - Range of ports to scan
   * @param {Array} range - An array of hostnames/ipv4/ipv6, CIDR or ranges
   * @param {Number} timeout - Number of seconds to wait for host/port response
   * @param {Number} blocksize - Number of hosts per network scanning block
   * @param {Number} threshold - Max number of  spawned process
   * @param {Array} flags - Array of flags for .spawn()
   * @param {Boolean} udp - Perform a scan using the UDP protocol
   * @param {Boolean} json - JSON object as output, false produces XML
   */
  const defaults = {
    nmap:       'nmap',
    verbose:    false,
    ports:      '1-1024',
    range: [],
    timeout:    120,
    blocksize:  16,
    threshold:  os.cpus().length * 4,
    flags: [
      '-T4',    // Scan optimization
    ],
    udp:        false,
    json:       true
  };

  /**
   * @method config
   * @abstract Configuration object
   */
  const config = {

    /**
     * @function init
     * @scope private
     * @abstract Merges supplied options & builds functions
     *
     * @param {Object} defaults libnmap default options
     * @param {Object} opts User supplied configuration object
     * @param {Function} cb Callback
     *
     * @returns {Object}
     */
    init(defaults, opts, cb) {
      const funcs = [];
      const called = caller.getData().functionName;

      /* Override 'defaults.flags' array with 'opts.flags' (prevents merge) */
      if (/array/.test(typeof opts.flags))
        defaults.flags = opts.flags;

      opts = tools.merge(defaults, opts);
      opts.called = called;

      /* Ensure we can parse the report */
      if (opts.flags.indexOf('-oX -') === -1)
        opts.flags.push('-oX -');

      validation.init(opts, (err, result) => {
        if (err)
          return cb(err);

        if (/discover/.test(called)) {
          if (!(opts.range = tools.adapters(opts)))
            return cb(new Error(validation.verErr));

          /* Set scan options as static values for 'discover' mode */
          opts.ports = '';
          opts.flags = [
            '-oX -',
            '-sn',
            '-PR'
          ];
        }

        opts.range = network.calculate(opts);
        funcs = tools.funcs(opts);

        return cb(null, {
          opts,
          funcs
        });
      });
    }
  };

  /**
   * @method reporting
   * @abstract Reporting object
   */
  const reporting = {

    /**
     * @function reports
     * @abstract Handle results
     *
     * @param {Obect} opts Application defaults
     * @param {Function} cb Return function
     *
     * @returns {Function}
     */
    reports(opts, report, cb) {
      if ((!/object/.test(typeof report)) || (report.hasOwnProperty('code')))
        return cb(new Error(report));

      const xml = report.join('');

      if (!opts.json)
        return cb(null, xml);

      try {
        const parserOptions = {
          attrkey: "item",
        };

        const xmlParser = new xml2js.Parser(parserOptions);

        xmlParser.parseString(xml, function parseXML(err, json) {
          if(err)
            return cb(new Error(err));

          cb(null, json.nmaprun);
        });
      } catch(err) {
        return cb(new Error(err));
      }
    }
  };

  /**
   * @method tools
   * @abstract Tools object
   */
  var tools = {

    /**
     * @function merge
     * @abstract Perform preliminary option/default object merge
     *
     * @param {Object} defaults Application defaults
     * @param {Object} obj User supplied object
     *
     * @returns {Object}
     */
    merge(defaults, obj) {
      return merge(defaults, obj);
    },

    /**
     * @function adapters
     * @abstract Obtain network adapter information and return an array of
     *           ranges as an array for CIDR calculations
     *
     * @param {Object} obj User supplied object
     *
     * @returns {Array}
     */
    adapters(obj) {
      const ret = [];
      const adapter = '';
      const netmask = '';
      const adapters = os.networkInterfaces();

      for (const iface in adapters) {

        for (const dev in adapters[iface]) {
          adapter = adapters[iface][dev];

          if (!adapter.internal) {

            if (!adapter.netmask)
              return false;

            if (adapter.netmask) {

              netmask = adapter.netmask;

              /* Convert netmask to CIDR notation if IPv6 */
              if (validation.test(validation.net.IPv6, netmask)) {
                netmask = new v6(netmask).subnet.substring(1);
              }

              ret.push(`${adapter.address}/${netmask}`);
            }
          }
        }
      }

      return ret;
    },

    /**
     * @function funcs
     * @abstract Create functions for use as callbacks
     *
     * @param {Obect} opts Application defaults
     *
     * @returns {Array}
     */
    funcs(opts) {
      const funcs = {};
      const cmd = false;
      const errors = [];
      const reports = [];

      if (opts.range > 0)
        return new Error("Range of hosts could not be created");

      Object.keys(opts.range).forEach(function blocks(block) {

        const range = opts.range[block];

        funcs[range] = function block(callback) {
          cmd = tools.command(opts, range);

          if (opts.verbose)
            console.log(`Running: ${cmd}`);

          const report = [];

          const execute = proc(cmd, function exe(err, stdout, stderr) {
              if (err)
                return reporting.reports(opts, err, callback);
            });

          execute.stderr.on('data', function errbytes(chunk) {
            /* Silently discard stderr messages to not interupt scans */
          });

          execute.stdout.on('data', function bytes(chunk) {
            report.push(chunk);
          });

          execute.stdout.on('end', function bytesend() {
            if (report.length > 0)
              return reporting.reports(opts, report, callback);
          });
        };
      });

      return funcs;
    },

    /**
     * @function command
     * @abstract Generate nmap command string
     *
     * @param {Object} opts - User supplied options
     * @param {String} block - Network block
     *
     * @returns {String} NMAP scan string
     */
    command(opts, block) {
      const flags = opts.flags.join(' ');
      const ip = new v6(block);
      const ipv6 = (ip.isValid()) ? ' -6 ' : ' ';
      const proto = (opts.udp) ? ' -sU' : ' ';
      const to = `--host-timeout=${opts.timeout}s `;

      return (opts.ports) ?
        `${opts.nmap+proto} ${to}${flags}${ipv6}-p${opts.ports} ${block}` :
        `${opts.nmap+proto} ${to}${flags}${ipv6}${block}`;
    },

    /**
     * @function worker
     * @abstract Executes object of functions
     *
     * @param {Object} obj User supplied object
     * @param {Function} fn Return function
     */
    worker(obj, fn) {
      async.parallelLimit(obj.funcs, obj.threshold, fn);
    }
  };

  /**
   * @method network
   * @abstract Network object
   */
  var network = {

    /**
     * @function range
     * @abstract Calculates all possible hosts per CIDR
     *
     * @param {Object} opts Application defaults
     * @param {Object} cidr - netmask module object
     *
     * @returns {Array}
     */
    range(opts, cidr) {
      const split = cidr.size / opts.blocksize;
      const results = [];

      split = (split > 256) ? Math.round(split / 255) : split;

      cidr.forEach((ip, long, index) => {

        if (index % split === 0) {
          const obj = ip.split('.');
          const start = parseInt(obj[3]);
          const end = (parseInt(obj[3]) + split) - 1;
          const range = null;

          range = (start === end) ?
            `${obj[0]}.${obj[1]}.${obj[2]}.${start}` :
            `${obj[0]}.${obj[1]}.${obj[2]}.${start}-${end > 255 ? 255 : end}`;

          results.push(range);
        }
      });

      return results;
    },

    /**
     * @function calculate
     * @abstract Performs calculation on subnet blocks
     *
     * @param {Object} opts Application defaults
     *
     * @returns {Array}
     */
    calculate(opts) {
      const blocks = [];
      const results = [];
      const cidr = false;
      const tests = validation.net;

      opts.range.forEach(host => {
        switch (true) {

          case (validation.test(tests.hostname, host) ||
                validation.test(tests.IPv4, host) ||
                validation.test(tests.IPv6, host)):

            results.push(host);

            break;

          case (validation.test(tests.IPv4CIDR, host) ||
                validation.test(tests.IPv6CIDR, host)):

            cidr = new netmask(host);
            blocks = network.range(opts, cidr);

            blocks.forEach(block => {
              results.push(block);
            });

            break;

          case (validation.test(tests.IPv4Range, host)):

            results.push(host);

            break;
          default:
            /* Silently discard specified element as invalid */
            break;
        }
      });

      return results;
    }
  };


  /**
   * @method validation
   * @abstract Validation object
   */
  var validation = {

    verErr: 'Discover method requires nodejs v0.11.2 or greater',

    pathErr: 'Supplied path for nmap binary is invalid',

    blockErr: 'Supplied blocksize must not exceed 128',

    rangeErr: 'Range must be an array of host(s). Examples: ' +
      '192.168.2.10 (single), 10.0.2.0/24 (CIDR), 10.0.10.5-20 (range)',

    portErr: 'Port(s) must match one of the following examples: ' +
      '512 (single) | 0-65535 (range) | 10-30,80,443,3306-10000 (multiple)',

    /**
     * @var net
     * @abstract Object with various REGEX patterns to validate network params
     */
    net: {

      /**
       * @var ports
       * @abstract Regex for matching port ranges
       * @ref http://stackoverflow.com/a/21075138/901697
       */
      ports: /^(?:(?:^|[-,])(?:[1-9][0-9]{0,3}|[1-5][0-9]{4}|6(?:[0-4][0-9]{3}|5(?:[0-4][0-9]{2}|5(?:[0-2][0-9]|3[0-5])))))+$/,

      /**
       * @var hostname
       * @abstract Regex for matching hostnames (RFC-1123)
       */
      hostname: /^(([a-zA-Z]|[a-zA-Z][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z]|[A-Za-z][A-Za-z0-9\-]*[A-Za-z0-9])|localhost$/,

      /**
       * @var IPv4
       * @abstract Regex for matching IPv4 address types
       */
      IPv4: /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/,

      /**
       * @var IPv4CIDR
       * @abstract Regex for matching IPv4 CIDR notation
       */
      IPv4CIDR: /(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(\/([1-2]\d|3[0-2]|\d))/,

      /**
       * @var IPv4Range
       * @abstract Regex for matching IPv4 Range notation
       */
      IPv4Range: /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\-([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/,

      /**
       * @var IPv6
       * @abstract Regex for matching IPv6 address types
       */
      IPv6: /^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*/,

      /**
       * @var IPv6CIDR
       * @abstract Regex for matching IPv6 CIDR notation
       */
      IPv6CIDR: /^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*(\/(\d|\d\d|1[0-1]\d|12[0-8]))$/,
    },

    /**
     * @function init
     * @abstract Construct for network/port validation
     *
     * @param {Object} opts - User supplied options
     * @param {Function} cb - Callback
     */
    init(opts, cb) {
      const scope = this;
      const errors = [];

      if (!scope.exists(opts.nmap))
        errors.push(new Error(scope.pathErr));

      if (opts.blocksize > 128)
        errors.push(new Error(scope.blockErr));

      if (!/discover/.test(opts.called)) {
        if ((!opts.range) || (!/array|object/.test(typeof(opts.range))) ||
            (opts.range.length === 0))
          errors.push(new Error(scope.rangeErr));

        if (opts.range.length >= 1) {
          opts.range.forEach(value => {
            scope.verify(value, (err, result) => {
              if (err) return errors.push(err);
            });
          });
        }
      }

      if (opts.ports) {
        if (!scope.net.ports.test(opts.ports))
          errors.push(new Error(scope.portErr));
      }

      return (errors.length > 0) ? cb(errors) : cb(null, true);
    },

    /**
     * @function verify
     * @abstract Verify options provided
     *
     * @param {String} host User supplied configuration object
     * @param {Function} cb - Callback
     *
     * @returns {Function}
     */
    verify(host, cb) {
      if (this.test(this.net.hostname, host) ||
          this.test(this.net.IPv4, host) ||
          this.test(this.net.IPv6, host) ||
          this.test(this.net.IPv4CIDR, host) ||
          this.test(this.net.IPv6CIDR, host) ||
				  this.test(this.net.IPv4Range, host)) {
        return cb(null, true);
      } else {
        return cb(new Error(`Supplied host (${host}) did not pass validation. ${this.rangeErr}`));
      }
    },

    /**
     * @function test
     * @abstract Test specified regex test on string
     *
     * @param {Object} regex - Regex test case
     * @param {String} str - String to perform test on
     *
     * @returns {Boolean}
     */
    test(regex, str) {
      return regex.test(str);
    },

    /**
     * @function exists
     * @abstract Binary file tests
     *
     * @param {String} path - Path for file
     *
     * @returns {Boolean}
     */
    exists(path) {
      return hasbin.sync(path);
    },
  };

  /**
   * @function discover
   * @abstract Finds online neighbors
   *
   * @param {Object} obj User supplied options
   * @param {Function} cb User supplied callback function
   */
  nmap.prototype.discover = (obj, cb) => {
    cb = cb || obj;

    const opts = {};

    config.init(defaults, obj, function config(err, settings) {
      if (err)
        return cb(err);

      opts = settings.opts;
      opts.funcs = settings.funcs;

      tools.worker(opts, function discover(err, data) {
        if (err)
          return cb(err);

        return cb(null, data);
      });
    });
  };

  /**
   * @function scan
   * @abstract Performs scan of specified host/port combination
   *
   * @param {Object} obj User supplied options
   * @param {Function} cb User supplied callback function
   */
  nmap.prototype.scan = (obj, cb) => {
    cb = cb || obj;

    const opts = {};

    config.init(defaults, obj, function config(err, settings) {
      if (err)
        return cb(err);

      opts = settings.opts;
      opts.funcs = settings.funcs;

      tools.worker(opts, function scan(err, data) {
        if (err)
          return cb(err);

        return cb(null, data);
      });
    });
  };
};

/* robot, do work */
module.exports = new nmap();